### PR TITLE
Enable HNC leader election in controllers

### PR DESCRIPTION
**Pull request by joe2far**
- updates hnc reconcilers to be leader aware and only perform writes when leader
- non leader instance run updating it's forest tree cache, but will not perform any writes
- all hnc instances can respond to validating / mutating webhooks
- when a leader election occurs objects are re-queued
- adds leases permissions to cluster role

---
manual testing completed with 3 instances and --enable-leader-election
- manually killing leader instance at different times and running tests creating many parent child ns hierarchys
- validated that the new instance took over and handled updating child ns resources 
- also validated that all 3 instances were successfully handling validation webhook calls

### CODE REVIEW

**Comment by adrianludwin** [comment]
Hey I'll review this shortly, but FYI I didn't know anyone had any big changes under way and moved a lot of files around this weekend :( Sorry about that! See #92 for details, but the summary is that instead of arranging the code by _function_ (reconcilers, mutators, and validators), I've arranged the code by _feature_ (anchors, hierarchyconfig, hncconfig, namespace, objects). Most of the old files still exist but have been renamed, e.g.:

```
internal/reconcilers/anchor.go -> internal/anchor/reconciler.go
internal/reconcilers/hierarchy_config.go -> internal/hierarchyconfig/reconciler.go
internal/reconcilers/hnc_config.go -> internal/hncconfig/reconciler.go
internal/reconcilers/object.go -> internal/objects/reconciler.go
internal/reconcilers/setup.go -> internal/setup/reconcilers.go
```

Hopefully most of the changes should be easy to port.

**Comment by joe2far** [comment]
@adrianludwin sorry, haven't had time to progress this further, it would be nice to have this in the upcoming release, but it won't be ready in time, I hope to get this PR in better state for review by mid next week

**Comment by adrianludwin** [comment]
Thanks, I think we're getting much closer now. Can you please update your commit message with more information about how you tested this change for posterity?

Also, I assume you got the reconciler setup code from somewhere - or did you just figure that out? If you got it from somewhere, it might be nice to mention that in the commit message as well.

**Comment by joe2far** [comment]
figured out the needed reconciler updates from review the logic in https://github.com/kubernetes-sigs/controller-runtime/tree/master/pkg specifically at https://github.com/kubernetes-sigs/hierarchical-namespaces/blob/master/vendor/sigs.k8s.io/controller-runtime/pkg/manager/internal.go#L209
 
there are some other controller-runtime consumers doing similar to move controller to nonLeaderElectionRunnables but it not wide spread and didn't find any specific kubebuilder project

testing was manual, with 3 instances and --enable-leader-election
manually killing leader instance at different times and load test creating many parent child ns hierarchys
validated that the new instance took over and handled updating child ns resources 
also validated that 3 instances were successfully handling validation webhook calls

**Comment by erikgb** [comment]
@joe2far What is the status on this? I think enabling HNC leader election would be excellent to get into the next release (GA, 1.0?). Please let me know if I can assist in any way to get this over the finish line. 😄

**Comment by erikgb** [comment]
I am not sure if introducing an additional dependency would be a good idea, but we might have a look at the "Leader for life" approach implemented as part of operator-sdk/operator-lib:

https://github.com/operator-framework/operator-lib/blob/0449c0c509287014f1095739b06279e9113105da/leader/leader.go#L88

Docs: https://sdk.operatorframework.io/docs/building-operators/golang/advanced-topics/#leader-for-life

**Comment by joe2far** [comment]
hi @erikgb , I have not looked at this for some time, so PR is stale but leader election behaviour was manually tested, when I was last looking into this....

the solution here, uses the leader election from `controller-runtime` which is k8s lease based, however in order for hnc reconcilers to work in leader-follower mode using this, they need to be updated with `NeedLeaderElection=false` to make them start as `NonLeaderElectionRunnables` and then based upon mgr.Elected() chan they can become leader instance ( only the leader instance  performs writes, whereas follower instances don't do any k8s writes but respond to webhooks and perform other actions, in order to keep their forest cache in sync, so they are warm instances, that can become the leader when elected)

I feel using `k8s-runtime` leader election, but setting up all mgr reconcilers as non leader election runnables is a bit hacky and would agree that using  a solution similar to leader election from https://github.com/operator-framework/operator-lib/ 
 may be better than, working around the the k8s-runtime, which is really more designed for  just a single active leader. instance.
@adrianludwin what are your thoughts on swapping from using `k8s-runtime` leader election, to use operator-framework/operator-lib/  leader or a similar implementation?

I will have some free cycles in the coming week and will look into reviving this PR (but would @erikgb welcome any assistance, on offer) and will probably swap over the leader election implementation used in this change.

**Comment by joe2far** [comment]
for example following is an example leader-follower setup,  using the operator-framework/operator-lib/ leader used by vault
see: https://github.com/hashicorp/vault-k8s/blob/c02054a37dd6c141c02975679a08cfa1053d245d/leader/leader.go

**Comment by erikgb** [comment]
One thing I am not able to understand: Why do we need a leader-follower leader-election? I would prefer if we could use the standard single leader approach provided by default by controller-runtime. We use that for all our in-house controllers, and can scale the number of replicas solely based on webhook requirements. The webhooks must work at all times, as it is extremely important for the general cluster availability, and I have heard people setting up 9 replicas just to reduce the possibility of no webhook endpoint not available.

I tested the standard controller-runtime leader election locally with HNC, and that does not work at all. And I'll guess that the reason this PR is suggesting to build some kind of custom leader election mechanism. Why wouldn't it be possible to use the standard leader election with some refactoring/changes? I think the webhooks should be able to work independently of the reconciler manager. So if you spin up 10 replicas, you will have 10 fully functional webhooks, but just a single manager performing reconciles. AFAIK most in-tree reconcilers in K8s runs with a single worker, i.e. ReplicaSet->Pod.

Maybe @adrianludwin can explain why HNC cannot do this? IMO the webhooks should have a read-only cache of the forest, while the reconcilers should have a read-write cache. Would that be possible to achieve? Where is the potential blocker? The webhook cert-issuer?

**Comment by adrianludwin** [comment]
Webhook performance has never been an issue since we serve entirely from
memory. I believe we can handle a *far* higher QPS than most K8s apiservers
can manage even with a single replica. However, even if this were not true,
my understanding is that the built-in controller-runtime leader election
ensures that only one replica runs the controller-manager. Of course, this
won't work for HNC, since it's the controller-manager (i.e. the
reconcilers) that maintain the state of the forest. So we'd need to
override the default behaviour and start all the reconcilers in all the
replicas, but ensure that only one of them is able to write.

While I think this is unnecessary, what _is_ necessary for HNC is to have a
replica be able to take over _quickly_ from the leader if it's lost, and
this requires exactly the same thing: the reconcilers must all be running
in a read-only mode, and then must switch to read-write mode as quickly as
possible (and also reconcile any changes that happened after the leader was
lost but before the election concluded). This is pretty nonstandard
behaviour AIUI.

I hope that explains the complexities here.

**Comment by erikgb** [comment]
I am not worried about webhook performance, but webhook **availability** is essential IMO. A failing webhook endpoint can affect all API calls in the cluster, so you typically want to run **at least** two replicas serving webhook endpoints - to be able to handle smooth updates and eventual cluster issues - like loosing a node. And several controllers I am aware of, run webhooks in dedicated workloads - just for this reason. One example is cert-manager. And this was also noted/suggested in this comment: https://github.com/kubernetes-sigs/hierarchical-namespaces/issues/68#issuecomment-978151389

Why cannot the webhooks run with their dedicated (read-only) forest - based on code shared with the reconcilers? I understand that this would eventually be a huge change in HNC, but HA for webhooks is important, and as a bonus we might be able to use the standard leader-election in controller-runtime.

When is the next community meeting? Maybe this subject is better addressed there, than as comments in this PR. 😉

**Comment by adrianludwin** [comment]
Good point about availability. Although at least the _object_ webhook fails
open IIRC so your cluster doesn't come to a crashing halt. But better if we
don't need to rely on that.

I'm not sure how you'd share code with the webhooks without basically
reimplementing _almost_ the entire reconciler. Most of the reconcilers'
logic is in fact about updating the forest to match what's on the server,
so it seems much easier to me to modify the reconcilers to run in read-only
mode (i.e., exactly what's necessary to create a read-only forest) than to
try to factor out the common code, and then set up our own custom
replacement for the controller-manager, just so that we can use standard
leader election. *Something's* going to have to be nonstandard :)

We don't usually discuss issues like this in our regular meetings, if
needed we could have a sync in the next couple of days but otherwise I'm ok
to keep this on PRs.

**Comment by erikgb** [comment]
> Good point about availability. Although at least the _object_ webhook fails open IIRC so your cluster doesn't come to a crashing halt. But better if we don't need to rely on that.

I would consider any webhook failing for "technical reasons" something that can be improved (bug). 😉 

> I'm not sure how you'd share code with the webhooks without basically reimplementing _almost_ the entire reconciler. Most of the reconcilers' logic is in fact about updating the forest to match what's on the server, so it seems much easier to me to modify the reconcilers to run in read-only mode (i.e., exactly what's necessary to create a read-only forest) than to try to factor out the common code, and then set up our own custom replacement for the controller-manager, just so that we can use standard leader election. *Something's* going to have to be nonstandard :)

As I wrote in my previous comment, I understand that changing this to something that represents reusable code will be a BIG refactoring. That said, I think it could make sense to decouple the forest cache and reconcilers in long-term. Somehow I imagine that the forest cache could include an informer component, that would be responsible for dispatching events to the reconcilers. And function as an adapter around the standard write-through cache + informer in controller-runtime.

> We don't usually discuss issues like this in our regular meetings, if needed we could have a sync in the next couple of days but otherwise I'm ok to keep this on PRs.

Yes, I would be interested in a sync - if you are available. Just ping me on Slack.

To summarize my stand on short-term goals, I think **enabling HA on webhooks is very important**. And to get it, it seems like we need some kind of leader election.

**Comment by erikgb** [comment]
> (leader election is a pretty simple feature)

I disagree 😉, and I think I have some support for my stand from controller-runtime and operator-lib/operator-sdk: https://sdk.operatorframework.io/docs/building-operators/golang/advanced-topics/#leader-election and https://github.com/kubernetes/client-go/blob/30b06a83d67458700a5378239df6b96948cb9160/tools/leaderelection/leaderelection.go#L21-L24.

> I'll try to give you a call on Friday morning Toronto time

Ack! 👍

**Comment by adrianludwin** [comment]
*Implementing* leader election is crazy hard, but I don't think that
controller-runtime is *using* it in any particularly special way. E.g. the
warnings on that comment about RenewDeadline and LeaseDuration are just
passed on to  the controller-runtime clients anyway (here
<https://github.com/kubernetes-sigs/controller-runtime/blob/f236f0345ad2933912ebf34bfcf0f93620769654/pkg/manager/manager.go#L178>).
At worst, we just use their sensible defaults (15s and 10s, respectively).

However, I doubt it would come to that. I'd probably do something like:

   - Make all reconcilers *not* require leader election; by default, they
   all run in read-only mode
   - Add a single controller-runtime Runnable called something like
   BecomeLeader; when it starts, it changes the mode and asks all the
   reconcilers to fully re-sync themselves

That should preserve the vast majority of both the controller-runtime LE
infrastructure, as well as have a minimal impact on HNC.

**Comment by adrianludwin** [comment]
Oh man I forgot to hit "submit" on a review I wrote in early November. Sorry! Will re-review now.

**Review by adrianludwin — COMMENTED** [review]
Left a few comments. @joe2far , thanks so much for reviving this.

I'll try to summarize the conversation I had with @erikgb  on Friday. Basically, his top priority is making the _webhooks_ highly available, and _not_ the reconciler. Technically, this means we don't actually need LE for this one purpose; instead, we could:

* Have a deployment of a single pod for the reconciler which _doesn't_ serve webhooks
* Have another deployment of 3 pods that _only_ serve the webhooks

These might be exactly the same image, just running with a different command-line flag. And it would probably work just fine. The only problem would be that if the reconciler pod ever died, it could be a few minutes (on a large cluster) before it became active again. And in that time, the webhooks could show some undefined behaviour if someone was busy modifying the hierarchy.

However, why should we wait for a few minutes for the reconciler pod to restart, if all the webhook pods already have all the information needed to reconcile as well? Also, this could use up a _lot_ of the apiserver's bandwidth for a while. Instead, we can use LE to just turn one of the replicas into the leader, since it already has all its information loaded into memory and ready to go. Even re-syncing all its reconcilers is probably very fast since all the API objects will be in the controller-runtime cache. There might still be a short period where the webhooks are not fully in sync with what's on the apiserver but that will at least resolve itself far more quickly.

@erikgb , just to your specific comments:

* What Joe's implemented is essentially the same of the "leader for life" approach you suggested, at least as I understand it. I've even suggested using the `atomic.Value` approach from that code here as well.
* You mentioned that you "tested the standard controller-runtime leader election locally with HNC, and that does not work at all." Were you testing what's currently built into HNC (which is known not to work) or this change that Joe's proposing? I wouldn't expect anything to work until this change is merged.

@erikgb , if you're happy with this I'm ready to merge this change (after a few nits), especially if @joe2far can manually test it one last time. @rjbez17 , can you please have a look at this as well?

Thanks everyone and sorry for the delays here!

File: internal/anchor/reconciler.go [hunk 1]
```diff
@@ -415,9 +422,72 @@ func (r *Reconciler) SetupWithManager(mgr ctrl.Manager) error {
 			}},
 		}
 	}
-	return ctrl.NewControllerManagedBy(mgr).
```
[the remark this answers is not in the document]
Comment (adrianludwin) [reply internal/anchor/reconciler.go:418]: Oh never mind, I forgot about the distinction between Controller and Reconciler. Sorry.

File: vendor/github.com/open-policy-agent/cert-controller/pkg/rotator/rotator.go [hunk 2]
```diff
@@ -163,6 +163,11 @@ type CertRotator struct {
 	caNotInjected          chan struct{}
 }
 
+// NeedLeaderElection is false as this runnable can run in standby mode without leader lock
+func (cr *CertRotator) NeedLeaderElection() bool {
+	return false
+}
+
```
Comment (adrianludwin) [inline vendor/github.com/open-policy-agent/cert-controller/pkg/rotator/rotator.go:170]: This seems a bit weird - what happens if multiple instances try to create conflicting certs for the same webhook at the same time? I wonder if this will _normally_ work but not _always_. Do you know?

Otherwise, an alternative option here might be to leave the runnables as requiring leader election, but then provide a function for the followers to determine when it's safe to start. E.g. maybe modify `ensureCertsMounted()` so that it's a standalone function that you can call (or just factor it out so as to preserve the existing interface), and it simply blocks until the certs exist. wdyt?

Either way, we should add this to the main cert-controller as well. Can you please file an issue to do so? Thanks!

**Comment by adrianludwin** [comment]
I think that #163 will be our solution for HA support, not leader election. We'll also focus on greatly accelerating startup time in v1.1, further reducing the need for active/passive replicas. Since there's not a lot of activity on this PR, let's close this and continue work on #163 and its derivatives.

/close
